lib/std/mem.zig

```diff
@@ -3225,6 +3227,244 @@ pub fn replaceScalar(comptime T: type, slice: []T, needle: T, replacement: T) vo
+/// must also not be empty.
+pub fn replacementSizeMany(comptime T: type, input: []const T, replacement_pairs: []const ReplacementPair(T)) usize {
+    assert(replacement_pairs.len > 0);
+    for (replacement_pairs) |pair| {
+        assert(pair.ToReplace.len > 0);
+    }
+
+    var i: usize = 0;
+    var size: usize = input.len;
+    while (i < input.len) {
+        var found = false;
+        for (replacement_pairs) |pair| {
+            if (mem.startsWith(T, input[i..], pair.ToReplace)) {
+                size = size - pair.ToReplace.len + pair.Replacer.len;
+                i += pair.ToReplace.len;
+                found = true;
+                break;
+            }
+        }
+
+        if (found == false) {
+            i += 1;
+        }
+    }
+    return size;
+}
+
+test "replacementSizeMany" {
+   var replacementPairs = [_]ReplacementPair(u8){
+        ReplacementPair(u8){
+            .ToReplace = "All",
+            .Replacer = "None of",
+        },
+        ReplacementPair(u8){
+            .ToReplace = "base",
+            .Replacer = "Zig",
+        },
+    };
+    try testing.expect(replacementSizeMany(u8, "All your base are belong to us", &replacementPairs) == 33);
+
+    replacementPairs[0] = ReplacementPair(u8){ .ToReplace = "reading", .Replacer = "writing" };
+    replacementPairs[1] = ReplacementPair(u8){ .ToReplace = "over", .Replacer = "not" };
+
+    try testing.expect(replacementSizeMany(u8, "Favor reading code over writing code", &replacementPairs) == 35);
+
+    // Empty needle is not allowed but input may be empty.
+    try testing.expect(replacementSizeMany(u8, "", &replacementPairs) == 0);
+
+    // Adjacent replacements.
+    replacementPairs[0] = ReplacementPair(u8){ .ToReplace = "\\n", .Replacer = "\n" };
+    replacementPairs[1] = ReplacementPair(u8){ .ToReplace = "\\t", .Replacer = "\t" };
+
+    try testing.expect(replacementSizeMany(u8, "\\n\\t\\n", &replacementPairs) == 3);
+
+    // keywords in keywords
+    replacementPairs[0] = ReplacementPair(u8){ .ToReplace = "version", .Replacer = "0.0.0" };
+    replacementPairs[1] = ReplacementPair(u8){ .ToReplace = "ver", .Replacer = "0.0.0" };
+    // Always pick the first on the array.
+    try testing.expect(replacementSizeMany(u8, "version", &replacementPairs) == 5);
+
+    // keywords in replacors
+    replacementPairs[0] = ReplacementPair(u8){ .ToReplace = "version", .Replacer = "ver" };
+    replacementPairs[1] = ReplacementPair(u8){ .ToReplace = "ver", .Replacer = "version" };
+
+    try testing.expect(replacementSizeMany(u8, "version", &replacementPairs) == 3);
+}
+
+/// Replace multiple needles with their respective replacement as many times as possible, 
+/// writing to an output buffer which is assumed to be of appropriate size.
+/// Use `replacementSizeMany` to calculate an appropriate buffer size.
+/// The `replacement_pairs`array must not be empty, as well as each `replacement_pair.ToReplace`.
```

Comment:
This sounds like some words are missing on explanation.

If I want to be pedantic, I could ask myself, if the replacement itself gets replaced or fi the needle is forwarded to the end of the original string upon replacement to prevent infinite recursive replacement.